File: tests/netDxf.Conformance/RawHandleOperationsTests.cs
using netDxf;
using netDxf.Header;
using netDxf.Entities;
using netDxf.IO;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static void RegisterRawHandleOperationsTests()
    {
        foreach (DxfVersion version in HandleProfiles)
        foreach (bool binary in new[] { false,true })
        {
            DxfVersion v=version;bool b=binary;
            Run($"handles/remap/permutation/{v}/{b}",()=>HandlePermutation(v,b));
            Run($"handles/remap/fresh-target/{v}/{b}",()=>HandleFreshTarget(v,b));
            Run($"handles/closure/{v}/{b}",()=>HandleClosure(v,b));
        }
        foreach (DxfVersion version in SupportedVersions)
        foreach (bool binary in new[] { false, true })
        { DxfVersion v=version;bool b=binary;Run($"handles/remap/typed/{v}/{b}",()=>HandleRemapTyped(v,b)); }
        foreach(int failure in Enumerable.Range(0,12))
        { int f=failure;Run($"handles/remap/reject/{f}",()=>HandleRemapInvalid(f)); }
        Run("handles/operations/cancel-stale-and-budget",HandleOperationGuards);
        Run("handles/operations/closure-evidence",HandleClosureEvidence);
    }
    private static List<DxfTag> HandleOperationFixture(DxfVersion version)
    {
        return new List<DxfTag>{
            new(0,"SECTION"),new(2,"HEADER"),new(9,"$ACADVER"),new(1,HandleProfileName(version)),new(9,"$DWGCODEPAGE"),new(3,"ANSI_1252"),
            new(9,"$HANDSEED"),new(5,"31"),new(0,"ENDSEC"),
            new(0,"SECTION"),new(2,"OBJECTS"),
            new(0,"DICTIONARY"),new(5,"10"),new(330,"0"),new(100,"AcDbDictionary"),new(3,"Hard"),new(360,"20"),new(3,"Soft"),new(350,"30"),
            new(0,"XRECORD"),new(5,"20"),new(330,"10"),new(100,"AcDbXrecord"),new(280,(short)1),new(1,"literal 10 is application text"),
            new(0,"XRECORD"),new(5,"30"),new(330,"10"),new(100,"AcDbXrecord"),new(280,(short)1),new(1,"data"),new(0,"ENDSEC"),
            new(0,"SECTION"),new(2,"ENTITIES"),new(0,"LINE"),new(5,"A"),
            new(102,"{ACAD_REACTORS"),new(330,"30"),new(102,"}"),
            new(330,"10"),new(100,"AcDbEntity"),new(100,"AcDbLine"),
            new(10,1.0),new(20,2.0),new(30,3.0),new(11,4.0),new(21,5.0),new(31,6.0),
            new(340,"20"),new(330,"99"),new(320,"10"),new(1001,"REFS"),new(1005,"30"),new(0,"ENDSEC"),new(0,"EOF")
        };
    }
    private static void HandlePermutation(DxfVersion version,bool binary)
    {
        var source=HandleRoundTrip(HandleOperationFixture(version),binary);var index=DxfRawHandleIndex.Create(source);
        using var original=new MemoryStream();source.Save(original);
        var mapping=new Dictionary<string,string>{{"0010","20"},{"20","10"},{"30","A"},{"a","30"}};
        var edited=index.RemapHandles(mapping);var next=DxfRawHandleIndex.Create(edited);
        Equal("DICTIONARY",next.FindDefinitions("20").Single().Record.Name,"Simultaneous swap overwrote identity");
        Equal("LINE",next.FindDefinitions("30").Single().Record.Name,"Permutation lost entity");
        Equal(source.Tags.Count,edited.Tags.Count,"Remap modified tag count");
        foreach(var item in index.Occurrences)
        {
            string actual=(string)edited.Tags[item.TagIndex].Value;
            if(item.Role is DxfRawHandleRole.Identity||item.IsReference)
            {
                string expected=item.CanonicalHandle switch{"10"=>"20","20"=>"10","30"=>"A","A"=>"30",_=>item.Handle};
                Equal(expected,actual,"Reference did not follow simultaneous permutation");
            }
            else Equal(item.Handle,actual,"Opaque/arbitrary/seed field was rewritten");
        }
        Check(source.HasOriginalBytes&&!edited.HasOriginalBytes,"Edited/original byte ownership");
        using var originalAgain=new MemoryStream();source.Save(originalAgain);
        Check(original.ToArray().SequenceEqual(originalAgain.ToArray()),"Source changed during remapping");
        using var output=new MemoryStream();edited.Save(output,!binary);output.Position=0;
        var loaded=DxfRawDocument.Load(output);SameRawTags(edited.Tags,loaded.Tags);
        var noop=next.RemapHandles(new Dictionary<string,string>{{"20","0020"}});
        Check(ReferenceEquals(noop,edited),"Numeric no-op allocated a changed snapshot");
        Throws<ArgumentException>(()=>next.GetOccurrences(index.FindDefinitions("10").Single().Record));
    }
    private static void HandleFreshTarget(DxfVersion version,bool binary)
    {
        var source=HandleRoundTrip(HandleOperationFixture(version),binary);var index=DxfRawHandleIndex.Create(source);
        var edited=index.RemapHandles(new Dictionary<string,string>{{"10","1000"}});var next=DxfRawHandleIndex.Create(edited);
        Equal("1001",next.Occurrences.Single(x=>x.Role==DxfRawHandleRole.HeaderSeed).CanonicalHandle,"HANDSEED did not advance");
        Equal(0,next.FindDefinitions("10").Count,"Old identity remained");
        Equal(1,next.FindDefinitions("1000").Count,"New identity absent");
        var changed=new HashSet<int>(index.Occurrences.Where(x=>(x.CanonicalHandle=="10"&&(x.IsReference||x.Role==DxfRawHandleRole.Identity))||x.Role==DxfRawHandleRole.HeaderSeed).Select(x=>x.TagIndex));
        for(int i=0;i<source.Tags.Count;i++)if(!changed.Contains(i))Check(ReferenceEquals(source.Tags[i],edited.Tags[i]),"Unrelated tag object replaced");
        Equal("10",next.Occurrences.Single(x=>x.Role==DxfRawHandleRole.Arbitrary).Handle,"Arbitrary handle was translated");
    }
    private static void HandleClosure(DxfVersion version,bool binary)
    {
        var index=DxfRawHandleIndex.Create(HandleRoundTrip(HandleOperationFixture(version),binary));
        var root=index.FindDefinitions("10").Single().Record;
        var closure=index.GetDependencyClosure(new[]{root,root});
        Equal(2,closure.Records.Count,"Default hard-ownership closure");Check(closure.AreSelectedReferencesResolved,"Hard closure unresolved");
        var all=index.GetDependencyClosure(new[]{root},DxfRawReferenceTraversal.All);
        Equal(3,all.Records.Count,"Soft ownership/parent closure");
        var line=index.FindDefinitions("A").Single().Record;
        var lineClosure=index.GetDependencyClosure(new[]{line},DxfRawReferenceTraversal.All);
        Equal(4,lineClosure.Records.Count,"Cycle-tolerant all-reference closure");
        Equal("99",lineClosure.UnresolvedReferences.Single().CanonicalHandle,"Dangling reference hidden");
        Check(!lineClosure.AreSelectedReferencesResolved,"Unresolved closure incorrectly certified");
        Equal(1,index.GetDependencyClosure(new[]{line},DxfRawReferenceTraversal.None).Records.Count,"No-traversal root identity");
        Throws<NotSupportedException>(()=>((IList<DxfRawRecord>)closure.Records).Clear());
    }
    private static void HandleRemapInvalid(int failure)
    {
        var tags=HandleOperationFixture(DxfVersion.AutoCad2018);
        var map=new Dictionary<string,string>{{"10","100"}};
        switch(failure)
        {
            case 0:map=new(){{"10","30"}};break; // unchanged identity collision
            case 1:map=new(){{"10","99"}};break; // capture a formerly dangling pointer
            case 2:map=new(){{"10","100"},{"20","100"}};break;
            case 3:map=new(){{"10","100"},{"0010","200"}};break;
            case 4:map=new(){{"0","100"}};break;
            case 5:map=new(){{"10","0"}};break;
            case 6:map=new(){{"99","100"}};break;
            case 7:tags.Insert(tags.FindIndex(t=>t.Code==1&&Equals(t.Value,"data")),new(330,"10"));break; // opaque XRECORD payload
            case 8:tags.InsertRange(tags.FindIndex(t=>t.Code==100&&Equals(t.Value,"AcDbEntity")),new DxfTag[]{new(102,"{VENDOR"),new(340,"100"),new(102,"}")});break;
            case 9:map=new(){{"10","FFFFFFFFFFFFFFFF"}};break;
            case 10:tags.InsertRange(tags.FindIndex(t=>t.Code==0&&Equals(t.Value,"LINE")),new DxfTag[]{new(0,"POINT"),new(5,"30")});break;
            case 11:tags.InsertRange(tags.FindIndex(t=>t.Code==9&&Equals(t.Value,"$HANDSEED")),new DxfTag[]{new(9,"$HANDSEED"),new(5,"31")});break;
        }
        var source=DxfRawDocument.Create(tags);var index=DxfRawHandleIndex.Create(source);
        if(failure<=6)Throws<ArgumentException>(()=>index.RemapHandles(map));
        else if(failure==9)Throws<OverflowException>(()=>index.RemapHandles(map));
        else Throws<InvalidOperationException>(()=>index.RemapHandles(map));
        for(int i=0;i<tags.Count;i++)Check(ReferenceEquals(tags[i],source.Tags[i]),"Failed remap mutated source");
    }
    private static void HandleOperationGuards()
    {
        var source=DxfRawDocument.Create(HandleOperationFixture(DxfVersion.AutoCad2018));var index=DxfRawHandleIndex.Create(source);
        using var cts=new CancellationTokenSource();cts.Cancel();
        Throws<OperationCanceledException>(()=>index.RemapHandles(new Dictionary<string,string>{{"10","100"}},cts.Token));
        Throws<OperationCanceledException>(()=>index.GetDependencyClosure(Array.Empty<DxfRawRecord>(),cancellationToken:cts.Token));
        Throws<ArgumentNullException>(()=>index.RemapHandles(null!));
        Throws<ArgumentNullException>(()=>index.GetDependencyClosure(null!));
        Throws<ArgumentOutOfRangeException>(()=>index.GetDependencyClosure(Array.Empty<DxfRawRecord>(),(DxfRawReferenceTraversal)256));
        var foreign=source.WithTags(source.Tags).Sections.SelectMany(s=>s.Records).First();
        Throws<ArgumentException>(()=>index.GetDependencyClosure(new[]{foreign}));
        bool disposed=false;
        IEnumerable<DxfRawRecord> Endless(){try{while(true)yield return index.FindDefinitions("10").Single().Record;}finally{disposed=true;}}
        var limited=DxfRawHandleIndex.Create(source,new DxfRawHandleIndexOptions(100,100));
        Throws<InvalidDataException>(()=>limited.GetDependencyClosure(Endless()));Check(disposed,"Bounded root enumerator leaked");
        Check(ReferenceEquals(source,index.RemapHandles(new Dictionary<string,string>())) ,"Empty mapping did not preserve original");
    }
    private static void HandleClosureEvidence()
    {
        var source=DxfRawDocument.Create(HandleFixture(DxfVersion.AutoCad2018));var index=DxfRawHandleIndex.Create(source);
        var closure=index.GetDependencyClosure(new[]{index.FindDefinitions("AB").Single().Record},DxfRawReferenceTraversal.All);
        Check(closure.UninterpretedHandles.Count>0,"Opaque closure limitation hidden");
        var tags=HandleOperationFixture(DxfVersion.AutoCad2018);int at=tags.FindIndex(t=>t.Code==0&&Equals(t.Value,"LINE"));
        tags.InsertRange(at,new DxfTag[]{new(0,"POINT"),new(5,"20")});
        var ambiguous=DxfRawHandleIndex.Create(DxfRawDocument.Create(tags));
        var result=ambiguous.GetDependencyClosure(new[]{ambiguous.FindDefinitions("10").Single().Record});
        Equal(1,result.AmbiguousReferences.Count,"Ambiguous targets silently selected");
        Equal(1,result.Records.Count,"Ambiguous dependency expanded");
    }
    private static void HandleRemapTyped(DxfVersion version, bool binary)
    {
        var typed=new DxfDocument(version); typed.Comments.Clear();
        typed.Entities.Add(new Line(new Vector3(1,2,3),new Vector3(4,5,6)));
        typed.Entities.Add(new Circle(new Vector3(7,8,9),2.5));
        using var stream=new MemoryStream();Check(typed.Save(stream,binary),"Remap typed fixture save");stream.Position=0;
        var raw=DxfRawDocument.Load(stream);var index=DxfRawHandleIndex.Create(raw);
        var map=index.Occurrences.Where(x=>x.Role==DxfRawHandleRole.Identity).ToDictionary(x=>x.CanonicalHandle,x=>(x.NumericHandle+4096).ToString("X"));
        var changed=index.RemapHandles(map);var after=DxfRawHandleIndex.Create(changed);
        Equal(index.Occurrences.Count,after.Occurrences.Count,"Full drawing remap changed handle inventory");
        Equal(index.Diagnostics.Count,after.Diagnostics.Count,"Full drawing remap introduced graph diagnostics");
        using var output=new MemoryStream();changed.Save(output,!binary);output.Position=0;
        var loaded=DxfDocument.Load(output)??throw new InvalidOperationException("Remapped drawing cannot load through typed reader");
        Equal(new Vector3(1,2,3),loaded.Entities.Lines.Single().StartPoint,"Remap changed line start");
        Equal(2.5,loaded.Entities.Circles.Single().Radius,"Remap changed circle radius");
        File.WriteAllBytes(Path.Combine(ArtifactDirectory,$"handle-remap-{version}-{binary}.dxf"),output.ToArray());
    }

}

File: javascript/tests/netDxf.Conformance/RawHandleOperationsTests.js
// Port of tests/netDxf.Conformance/RawHandleOperationsTests.cs; typed controls remain unported.
import { DxfTag, DxfRawDocument, DxfVersion, DxfRawHandleIndex, DxfRawHandleRole as R,
  DxfRawReferenceTraversal as F, DxfRawHandleIndexOptions } from '../../index.js';
import * as E from '../../runtime/Errors.js';
import { Run, Check, Equal, Throws, VersionName, BooleanName } from './TestHarness.js';
import { SaveRaw, LoadRaw, SameRawTags } from './RawDocumentTests.js';
import { HandleProfiles, HandleProfileName, HandleFixture, HandleRoundTrip } from './RawHandleIndexTests.js';
const T = (code, value) => new DxfTag(code, value);
export function RegisterRawHandleOperationsTests() {
  for (const v of HandleProfiles) for (const b of [false, true]) {
    Run(`handles/remap/permutation/${VersionName(v)}/${BooleanName(b)}`, () => HandlePermutation(v, b));
    Run(`handles/remap/fresh-target/${VersionName(v)}/${BooleanName(b)}`, () => HandleFreshTarget(v, b));
    Run(`handles/closure/${VersionName(v)}/${BooleanName(b)}`, () => HandleClosure(v, b));
  }
  for (let f = 0; f < 12; f++) Run(`handles/remap/reject/${f}`, () => HandleRemapInvalid(f));
  Run('handles/operations/cancel-stale-and-budget', HandleOperationGuards);
  Run('handles/operations/closure-evidence', HandleClosureEvidence);
}
export function HandleOperationFixture(version) {
  return [T(0,'SECTION'),T(2,'HEADER'),T(9,'$ACADVER'),T(1,HandleProfileName(version)),T(9,'$DWGCODEPAGE'),T(3,'ANSI_1252'),
    T(9,'$HANDSEED'),T(5,'31'),T(0,'ENDSEC'),T(0,'SECTION'),T(2,'OBJECTS'),
    T(0,'DICTIONARY'),T(5,'10'),T(330,'0'),T(100,'AcDbDictionary'),T(3,'Hard'),T(360,'20'),T(3,'Soft'),T(350,'30'),
    T(0,'XRECORD'),T(5,'20'),T(330,'10'),T(100,'AcDbXrecord'),T(280,1),T(1,'literal 10 is application text'),
    T(0,'XRECORD'),T(5,'30'),T(330,'10'),T(100,'AcDbXrecord'),T(280,1),T(1,'data'),T(0,'ENDSEC'),
    T(0,'SECTION'),T(2,'ENTITIES'),T(0,'LINE'),T(5,'A'),T(102,'{ACAD_REACTORS'),T(330,'30'),T(102,'}'),
    T(330,'10'),T(100,'AcDbEntity'),T(100,'AcDbLine'),T(10,1),T(20,2),T(30,3),T(11,4),T(21,5),T(31,6),
    T(340,'20'),T(330,'99'),T(320,'10'),T(1001,'REFS'),T(1005,'30'),T(0,'ENDSEC'),T(0,'EOF')];
}
export function HandlePermutation(v, b) {
  const source = HandleRoundTrip(HandleOperationFixture(v), b), index = DxfRawHandleIndex.Create(source), original = SaveRaw(source);
  const edited = index.RemapHandles(new Map([['0010','20'],['20','10'],['30','A'],['a','30']]));
  const next = DxfRawHandleIndex.Create(edited);
  Equal('DICTIONARY', next.FindDefinitions('20')[0].Record.Name); Equal('LINE', next.FindDefinitions('30')[0].Record.Name);
  Equal(source.Tags.Count, edited.Tags.Count);
  const changes = { '10':'20', '20':'10', '30':'A', 'A':'30' };
  for (const item of index.Occurrences) {
    const expected = item.Role === R.Identity || item.IsReference ? changes[item.CanonicalHandle] ?? item.Handle : item.Handle;
    Equal(expected, edited.Tags[item.TagIndex].Value);
  }
  Check(source.HasOriginalBytes && !edited.HasOriginalBytes); Equal(original, SaveRaw(source));
  SameRawTags(edited.Tags, LoadRaw(SaveRaw(edited, !b)).Tags);
  Check(next.RemapHandles(new Map([['20','0020']])) === edited);
  Throws(E.ArgumentException, () => next.GetOccurrences(index.FindDefinitions('10')[0].Record));
}
export function HandleFreshTarget(v, b) {
  const source = HandleRoundTrip(HandleOperationFixture(v), b), index = DxfRawHandleIndex.Create(source);
  const edited = index.RemapHandles(new Map([['10','1000']])), next = DxfRawHandleIndex.Create(edited);
  Equal('1001', next.Occurrences.find(x => x.Role === R.HeaderSeed).CanonicalHandle);
  Equal(0, next.FindDefinitions('10').Count); Equal(1, next.FindDefinitions('1000').Count);
  const changed = new Set(index.Occurrences.filter(x =>
    (x.CanonicalHandle === '10' && (x.IsReference || x.Role === R.Identity)) || x.Role === R.HeaderSeed).map(x => x.TagIndex));
  for (let i = 0; i < source.Tags.Count; i++) if (!changed.has(i)) Check(source.Tags[i] === edited.Tags[i]);
  Equal('10', next.Occurrences.find(x => x.Role === R.Arbitrary).Handle);
}
export function HandleClosure(v, b) {
  const index = DxfRawHandleIndex.Create(HandleRoundTrip(HandleOperationFixture(v), b)), root = index.FindDefinitions('10')[0].Record;
  const closure = index.GetDependencyClosure([root, root]);
  Equal(2, closure.Records.Count); Check(closure.AreSelectedReferencesResolved);
  Equal(3, index.GetDependencyClosure([root], F.All).Records.Count);
  const line = index.FindDefinitions('A')[0].Record, all = index.GetDependencyClosure([line], F.All);
  Equal(4, all.Records.Count); Equal(1, all.UnresolvedReferences.Count); Equal('99', all.UnresolvedReferences[0].CanonicalHandle);
  Check(!all.AreSelectedReferencesResolved); Equal(1, index.GetDependencyClosure([line], F.None).Records.Count);
  Throws(TypeError, () => closure.Records.splice(0));
}
export function HandleRemapInvalid(failure) {
  const tags = HandleOperationFixture(DxfVersion.AutoCad2018); let map = new Map([['10','100']]);
  switch (failure) {
    case 0: map = new Map([['10','30']]); break;
    case 1: map = new Map([['10','99']]); break;
    case 2: map = new Map([['10','100'],['20','100']]); break;
    case 3: map = new Map([['10','100'],['0010','200']]); break;
    case 4: map = new Map([['0','100']]); break;
    case 5: map = new Map([['10','0']]); break;
    case 6: map = new Map([['99','100']]); break;
    case 7: tags.splice(tags.findIndex(t => t.Code === 1 && t.Value === 'data'), 0, T(330,'10')); break;
    case 8: tags.splice(tags.findIndex(t => t.Code === 100 && t.Value === 'AcDbEntity'), 0, T(102,'{VENDOR'),T(340,'100'),T(102,'}')); break;
    case 9: map = new Map([['10','FFFFFFFFFFFFFFFF']]); break;
    case 10: tags.splice(tags.findIndex(t => t.Code === 0 && t.Value === 'LINE'), 0, T(0,'POINT'),T(5,'30')); break;
    case 11: tags.splice(tags.findIndex(t => t.Code === 9 && t.Value === '$HANDSEED'), 0, T(9,'$HANDSEED'),T(5,'31')); break;
  }
  const source = DxfRawDocument.Create(tags), index = DxfRawHandleIndex.Create(source);
  Throws(failure <= 6 ? E.ArgumentException : failure === 9 ? E.OverflowException : E.InvalidOperationException, () => index.RemapHandles(map));
  for (let i = 0; i < tags.length; i++) Check(tags[i] === source.Tags[i]);
}
export function HandleOperationGuards() {
  const source = DxfRawDocument.Create(HandleOperationFixture(DxfVersion.AutoCad2018)), index = DxfRawHandleIndex.Create(source);
  const cancelled = { aborted: true };
  Throws(E.OperationCanceledException, () => index.RemapHandles(new Map([['10','100']]), cancelled));
  Throws(E.OperationCanceledException, () => index.GetDependencyClosure([], undefined, cancelled));
  Throws(E.ArgumentNullException, () => index.RemapHandles(null)); Throws(E.ArgumentNullException, () => index.GetDependencyClosure(null));
  Throws(E.ArgumentOutOfRangeException, () => index.GetDependencyClosure([], 256));
  const foreign = source.WithTags(source.Tags).Sections.flatMap(s => s.Records)[0];
  Throws(E.ArgumentException, () => index.GetDependencyClosure([foreign]));
  let disposed = false;
  function* Endless() { try { while (true) yield index.FindDefinitions('10')[0].Record; } finally { disposed = true; } }
  const limited = DxfRawHandleIndex.Create(source, new DxfRawHandleIndexOptions(100, 100));
  Throws(E.InvalidDataException, () => limited.GetDependencyClosure(Endless())); Check(disposed);
  Check(source === index.RemapHandles(new Map()));
}
export function HandleClosureEvidence() {
  const source = DxfRawDocument.Create(HandleFixture(DxfVersion.AutoCad2018)), index = DxfRawHandleIndex.Create(source);
  const closure = index.GetDependencyClosure([index.FindDefinitions('AB')[0].Record], F.All); Check(closure.UninterpretedHandles.Count > 0);
  const tags = HandleOperationFixture(DxfVersion.AutoCad2018), at = tags.findIndex(t => t.Code === 0 && t.Value === 'LINE');
  tags.splice(at, 0, T(0,'POINT'),T(5,'20'));
  const ambiguous = DxfRawHandleIndex.Create(DxfRawDocument.Create(tags));
  const result = ambiguous.GetDependencyClosure([ambiguous.FindDefinitions('10')[0].Record]);
  Equal(1, result.AmbiguousReferences.Count); Equal(1, result.Records.Count);
}
